src/ripple/overlay/impl/ProtocolVersion.cpp

```diff
@@ -0,0 +1,142 @@
+
+#include <ripple/overlay/impl/ProtocolVersion.h>
+#include <ripple/beast/core/LexicalCast.h>
+#include <ripple/beast/rfc2616.h>
+#include <boost/regex.hpp>
+#include <algorithm>
+
+namespace ripple {
+
+/** The list of protocol versions we speak and we prefer to use. */
+constexpr
+ProtocolVersion const supportedProtocolList[]
+{
+    { 1, 2 },
+    { 2, 0 }
+};
+
+#if __cplusplus > 201703L
+static_assert(
+    std::is_sorted(supportedProtocolList.begin(), supportedProtocolList.end()),
+        "The list of supported protocols must be in sorted order.");
+#endif
+
+std::string
+to_string(ProtocolVersion const& p)
+{
+    // The legacy protocol uses a different name. This can be removed when we
+    // migrate away from it and require 2.0 or later.
+    if (p == { 1, 2 })
+        return "RTXP/1.2";
+
+    return "XRPL/" + std::to_string(p.first) + "." + std::to_string(p.second);
+}
+
+std::vector<ProtocolVersion>
+parseProtocolVersions(boost::beast::string_view const& value)
+{
+    static boost::regex re(
+        "^"                  // start of line
+        "(?:"                // Alternative #1: old-style
+        "RTXP/(1)\\.(2)"     // The string "RTXP/1.2"
+        "|"                  // Alternative #2: new-style
+        "XRPL/"              // the string "XRPL/"
+        "([2-9][0-9]*)"      // a number (greater than 2 with no leading zeroes)
+        "\\."                // a period
+        "(0|[1-9][0-9]*)"    // a number (no leading zeroes unless exactly zero)
+        ")"
+        "$"                  // The end of the string
+        , boost::regex_constants::optimize);
+
+    std::vector<ProtocolVersion> result;
+
+    for (auto const& s : beast::rfc2616::split_commas(value))
+    {
+        boost::smatch m;
+        if (boost::regex_match(s, m, re))
+        {
+            std::uint16_t major;
+            std::uint16_t minor;
+            if (!beast::lexicalCastChecked(major, std::string(m[1])))
+                continue;
+            if (!beast::lexicalCastChecked(minor, std::string(m[2])))
+                continue;
+            result.push_back(make_protocol(major, minor));
+        }
+    }
+    // We guarantee that the returned list is sorted and contains no duplicates:
+    std::sort(result.begin(), result.end());
+    result.erase(std::unique(result.begin(), result.end()), result.end());
+
+    return result;
+}
+
+boost::optional<ProtocolVersion>
+negotiateProtocolVersion(boost::beast::string_view const& versions)
+{
+    auto const them = parseProtocolVersions(versions);
+
+    std::vector<ProtocolVersion> common;
+    common.reserve(
+        std::distance(
+            std::begin(supportedProtocolList),
+            std::end(supportedProtocolList)));
+    std::set_intersection(
+        them.begin(), them.end(),
+        std::begin(supportedProtocolList),
+        std::end(supportedProtocolList),
+        std::back_inserter(common));
+
+    auto x = std::max_element(common.begin(), common.end());
```

Comment:
The largest element will always be on the end, no need to search.

I just thought of something "cute". First, two observations:
1) We don't need to allocate at all. If we use the same algorithm as `std::set_intersection`, use reverse iterators, the result is always the first element that would be written. `std::set_intersection` isn't set up to do that, and it isn't worth it to write a custom version.
2) If we use `set_intersection` as-is and provide a custom iterator that calls a function, we can get most of the same thing. We could write to an optional on every iteration. The result is the final value of the optional. Something like:
```

boost::optional<ProtocolVersion>
negotiateProtocolVersion(boost::beast::string_view const& versions)
{
    auto const them = parseProtocolVersions(versions);
    boost::optional<ProtocolVersion> result;
    std::set_intersection(
        them.begin(),
        them.end(),
        std::begin(supportedProtocolList),
        std::end(supportedProtocolList),
        boost::make_function_output_iterator(
            [&result](auto const& e) { result = e; }));
    return result;
}
```
(we need to include `#include <boost/function_output_iterator.hpp>` for that to work).

BTW, I'm 100% fine with the current code (my "cute" code takes a minute to decipher while yours is clear). Just wanted to share the thought.

Reply:
I actually like that a lot.